Feature/757 notification error queue (#76)
maykinmedia/notifications-api-common: docs/delivery_guarantees.rst

notification and forwards the message to the registered webhook.

.. _webhooks: https://en.wikipedia.org/wiki/Webhook

Failure modes
-------------

Even though the mechanism is simple, the underlying infrastructure is not. There is
always a chance that a message does not get properly delivered - a problem that all
*message broker* systems have.

The Notificaties API standard defines that recipients of a message/notification have to
reply with a HTTP 204 status code to confirm that the message was received. However,
to complicate things further, this confirmation response may also be lost. To summarize,
the following scenarios are possible:

* notifications_api_common delivers message and receives confirmation (happy flow)
* notifications_api_common delivers message but does not receive a confirmation (failure mode)
* notifications_api_common fails to deliver the message successfully (failure mode)

Now there are essentially two mitigation modes available:

* at-most-once delivery
* at-least-once delivery

Delivering a message exactly once is not possible since the underlying infrastructure
("the internet") may fail for whatever reason.

Retry mechanism
~~~~~~~~~~~~~~~

> By default, sending notifications to Open Notificaties has automatic retry behaviour, i.e. if the notification

Reviewer: ```suggestion
By default, sending notifications to Open Notificaties has automatic retry behaviour, i.e. if the notification or cloudevent
```

  delay in seconds between task autoretries. Default is ``52000`` seconds.
* **Notification delivery base factor**: the base factor used for exponential backoff.
  This can be increased or decreased to spread retries over a longer or shorter time period.
  Default is ``4``.

With the assumption that the requests are done immediately we can model the notification
tasks schedule with the default configurations:

1. At 0s the request to send a Notification to a subscriber is made, the notification task is scheduled, picked up
   by worker and failed
2. At 25s with 25s delay the first retry happens (``4^0`` * ``Notification delivery retry backoff``)
3. At 2m5s with 100s delay - the second retry (``4^1`` * ``Notification delivery retry backoff``)
4. At 8m45s with 400s delay - the third retry
5. At 35m25s with 1600s delay - the fourth retry
6. At 2h22m5s with 6400s delay - the fifth retry
7. At 9h28m45s with 25600s delay - the sixth retry
8. At 23h55m25s with 52000s delay - the seventh and final retry, capped by max delay.

So if the subscribed webhooks is up after 1 min of downtime the default configuration can handle it
automatically.

LOG_NOTIFICATIONS_IN_DB
~~~~~~~~~~~~~~~~~~~~~~~

> When ``LOG_NOTIFICATIONS_IN_DB`` is set to ``True``, failed notifications are stored in the database with all their failed requests.

Reviewer: ```suggestion
When ``LOG_NOTIFICATIONS_IN_DB`` is set to ``True``, failed notifications and failed cloud events are stored in the database with all their failed requests.
```

> From the admin they can be manually restarted. The Notification will be deleted on a successful request.

Reviewer: ```suggestion
From the admin they can be manually re-sent. The Notification will be deleted on a successful request.
```

> With the command ``clean_failed_notifications`` all notifications older than ``NOTIFICATION_NUMBER_OF_DAYS_RETAINED`` (default 60 days) can be removed.

Reviewer: Could you add a code-block to show how to run this command?

Author: added

   by worker and failed
2. At 25s with 25s delay the first retry happens (``4^0`` * ``Notification delivery retry backoff``)
3. At 2m5s with 100s delay - the second retry (``4^1`` * ``Notification delivery retry backoff``)
4. At 8m45s with 400s delay - the third retry
5. At 35m25s with 1600s delay - the fourth retry
6. At 2h22m5s with 6400s delay - the fifth retry
7. At 9h28m45s with 25600s delay - the sixth retry
8. At 23h55m25s with 52000s delay - the seventh and final retry, capped by max delay.

So if the subscribed webhooks is up after 1 min of downtime the default configuration can handle it
automatically.

LOG_NOTIFICATIONS_IN_DB
~~~~~~~~~~~~~~~~~~~~~~~
When ``LOG_NOTIFICATIONS_IN_DB`` is set to ``True``, failed notifications and failed cloud events are stored in the database with all their failed requests.
From the admin they can be manually re-sent. The Notification will be deleted on a successful request.

With the command ``clean_failed_notifications`` all notifications older than ``NOTIFICATION_NUMBER_OF_DAYS_RETAINED`` (default 60 days) can be removed.

.. code-block:: bash

>     python manage.py clean_failed_notifications

Reviewer: ```suggestion
    python src/manage.py clean_failed_notifications
```
